open_gpt/models/modeling.py

```diff
@@ -61,18 +69,35 @@ def load_model_and_transforms(
     ):
         from .loading import load_model_and_tokenizer
 
-        self.model, self.tokenizer = load_model_and_tokenizer(
-            model_name_or_path,
-            tokenizer_name_or_path=tokenizer_name_or_path,
-            precision=self._precision,
-            dtype=self._dtype,
-            device=self._device,
-            device_map=self._device_map,
-            use_fast=False,
-        )
-
-        if adapter_name_or_path is not None:
-            self.load_adapter(adapter_name_or_path)
+        if self._backend == 'hf':
+            self.model, self.tokenizer = load_model_and_tokenizer(
+                model_name_or_path,
+                tokenizer_name_or_path=tokenizer_name_or_path,
+                precision=self._precision,
+                dtype=self._dtype,
+                device=self._device,
+                device_map=self._device_map,
+                use_fast=False,
+            )
+
+            if adapter_name_or_path is not None:
+                self.load_adapter(adapter_name_or_path)
+        elif self._backend == 'vllm':
+            if not torch.cuda.is_available():
+                raise RuntimeError(f"vllm must be used with cuda.")
+            if self._precision not in ['fp32', 'fp16']:
+                raise ValueError(f"vllm only supports `float32`, `float16`, and `bfloat16`, got: {self._precision}")
+
+            from vllm import LLM
+
+            self.model = LLM(model=model_name_or_path,
+                             tokenizer=tokenizer_name_or_path,
+                             tokenizer_mode='slow',
+                             tensor_parallel_size=mself._tensor_parallel_size or torch.cuda.device_count(),
+                             # Pipeline parallelism is not supported yet.
+                             pipeline_parallel_size=1,
+                             trust_remote_code=True)
+            self.tokenizer = self.model.get_tokenizer()
```

Comment:
I guess it cannot work with LoRA finetuning as vllm has its own model implementation.